### DISCUSSION TITLE
Wondering how the RPC operation for the new queues works

### DISCUSSION THREAD

**Discussion by iLykaPaul** [supabase/supabase#31065] [Questions] [answered]
I've just been trying out the new (awesome) queues function and ran into an issue which I'm probably misunderstanding....

(my use case is currently server side API calling Supabase, not client side)

Following the documentation:

1. Enable extension
2. Expose Queues via PostgREST
3. Set RLS permissions
4. ... etc

The docs advise that "_Queues, by default, are not exposed over Supabase Data API and are only accessible via Postgres clients_." I believe this is enabled (in my instance) and I can see from the following that ```pgmq_public``` is listed. My understanding is that this exposes several functions such as ```send```.

![CleanShot 2024-12-11 at 20 28 13](https://github.com/user-attachments/assets/fc7c9825-a00d-4931-a17c-238345aeedd3)

![CleanShot 2024-12-11 at 20 22 26](https://github.com/user-attachments/assets/15b3005e-6a9d-46af-ada1-30bf6af5a5e6)

Taking this forward I then have an operation on my own API to trigger a job to be places on the queue:

```
const response = await Client.connection.rpc('send', {
      queue_name: 'diary_delete',
      message: { hello: 'world' },
      sleep_seconds: 30
})
```

(which is a near copy/paste from the docs)

However I am getting back the following error message:

"'_Could not find the function public.send(message, queue_name, sleep_seconds) in the schema cache'_"

The point I note here is that the RPC call is searching within the 'public' schema and not in ```pgmq_public```, which is of course where the 'send' function lives.

Having read through the docs numerous times I've started to wonder whether there is a need to "switch" the schema in advance of the RPC call or something similar? From what I can see it's all configured as per the instructions given, but my call to queue is failing.

I'm not suggesting a bug here at all, but I can't quite see whats missing for the 'send' function to be found.

Any help greatly appreciated.

**Comment by GaryAustin1** [comment, marked as answer]
You can set schema on individual calls with .schema()

**Reply by iLykaPaul** [reply]
Thank you @GaryAustin1, this got me past the error. I would humbly suggest that this should be put into the docs as I don't think this is clear for first time users.

I wonder too if you might provide me a steer on this line in the docs too:

> On top of enabling RLS and writing RLS policies on the underlying Queue tables, **you must grant the correct permissions to the pgmq_public database functions for each Data API role**.

I've read this sentence about 47 times and I have to see the clouds clear on what it's referring to. I can't see where this screen is from the docs either:

![CleanShot 2024-12-12 at 08 00 02](https://github.com/user-attachments/assets/acfc397f-2a77-4383-aa7c-72f56565bc97)

(apologies if this is all really obvious... )

**Reply by GaryAustin1** [reply]
If you feel the docs are missing something I highly recommend you generate a doc issue in Supabase/Supabase issues or if you are adventurous a PR with the change.  I’m not an employee so would have to do the same.

**Reply by GaryAustin1** [reply]
As far as grants you have to use SQL normally to grant execution access to non public schema functions.  The AI or Google should show you that.  I’ve not used these yet so don’t know if there is more involved.

**Reply by iLykaPaul** [reply]
Ah my bad @GaryAustin1 I made the assumption you were on the SB team given your activity in these discussion boards. Sorry about that.

Thanks for the steer, I'll do some research and feedback here once I've gotten further so others might benefit too.

:)

**Comment by iLykaPaul** [comment]
For reference the queue settings are obvious (now that I see cog): 

![CleanShot 2024-12-12 at 09 10 14](https://github.com/user-attachments/assets/4703dbfe-e9f7-4228-bfa4-927fcc589c88)
